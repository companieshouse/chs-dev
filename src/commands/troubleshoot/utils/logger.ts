import { simpleColouriser } from "../../../helpers/colouriser.js";

export default class TroubleshootOutputLogger {

    private readonly logger: (msg: string) => void;

    constructor (logger: (msg: string) => void) {
        this.logger = logger;
    }

    logMessage (message: string, stat: "Success" | "Error"): void {
        stat === "Success"
            ? this.logger(`${simpleColouriser("SUCCESS!", "bold-green")} ${message}`)
            : this.logger(`${simpleColouriser("FAILED!", "red")} ${message}`);

    }
}
