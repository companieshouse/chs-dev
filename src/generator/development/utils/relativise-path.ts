import { dirname, join, relative } from "path";

/**
 * Creates relative paths from the destination directory to the source for the
 * supplied paths supplied.
 * @param paths Array of strings or string representing the paths being
 *      converted to relative paths
 * @param source original location from which the path was relative to prior
 * @param destinationDirectory Location from which the new path will be
 *      relative from
 * @returns relative path from the destination directory to the paths provided
 */
export const relativiseServicePath = (
    paths: string | string[],
    source: string,
    destinationDirectory: string
) => {
    if (Array.isArray(paths)) {
        return paths.map(
            singleValue => relativiseServicePath(singleValue, source, destinationDirectory)
        );
    }

    const relativePathToSource = relative(destinationDirectory, source);

    return join(dirname(relativePathToSource), paths);
};

export default relativiseServicePath;
