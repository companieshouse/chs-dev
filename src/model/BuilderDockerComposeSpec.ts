/**
 * Represents a Docker Compose specification for a builder
 */
export interface BuilderDockerComposeSpec {
    /**
     * Name of the builder
     */
    name: string;

    /**
     * Version of the builder
     */
    version: string;

    /**
     * String representation of the docker compose spec - may contain variables
     * in '< >' which are removed when the docker compose file is created
     */
    builderSpec: string;
}

export default BuilderDockerComposeSpec;
