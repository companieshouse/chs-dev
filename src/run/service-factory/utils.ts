import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
