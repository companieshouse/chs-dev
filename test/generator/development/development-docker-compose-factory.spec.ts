import BuilderDockerComposeSpec from "../../../src/model/BuilderDockerComposeSpec";
import DevelopmentDockerComposeSpecFactory from "../../../src/generator/development/development-docker-compose-factory";

import { generateBuilderSpec, generateServiceSpec } from "../../utils/docker-compose-spec";
import { services } from "../../utils/data";
import { expect } from "@jest/globals";

const unmodifiableFieldTestCaseData = [
    // labels
    { labels: ["label.one=one", "label.two=two"] },
    {
        environment: {
            one: "one"
        }
    },
    {
        labels: {
            label: "one"
        },
        environment: [
            "one=one"
        ]
    }, {
        networks: ["chs"]
    }, {
        image: "123456.ecr.eu-west-2.amazonaws.com/repo"
    }, {
        image: "123456.ecr.eu-west-2.amazonaws.com/repo",
        labels: ["one=1", "two=2"],
        environment: {
            ENV_VAR: "yes"
        },
        networks: [
            "chs"
        ]
    },
    {
        healthcheck: {
            test: "echo hello"
        }
    }
];

describe("DevelopmentDockerComposeSpecFactory create", () => {

    const builderSpec: BuilderDockerComposeSpec = {
        name: "java",
        version: "v2",
        builderSpec: generateBuilderSpec("locals/builder/java/v2", true)
    };

    const projectPath = "/home/test-user/projects/docker-chs";

    let developmentDockerComposeSpecFactory: DevelopmentDockerComposeSpecFactory;

    beforeEach(() => {
        developmentDockerComposeSpecFactory = new DevelopmentDockerComposeSpecFactory(
            builderSpec, projectPath
        );
    });

    for (const testCase of unmodifiableFieldTestCaseData) {
        it(`adds unmodifiable field(s): ${Object.keys(testCase).join(", ")} to development service`, () => {
            const service = services[0];
            const serviceSpec = generateServiceSpec(service);

            serviceSpec.services[service.name] = {
                ...serviceSpec.services[service.name],
                ...testCase
            };

            const result = developmentDockerComposeSpecFactory.create(
                serviceSpec, service
            );

            for (const unmodifiableField of Object.keys(testCase)) {
                expect(result.services[service.name][unmodifiableField]).toBe(
                    testCase[unmodifiableField]
                );
            }
        });
    }

    it("adds builder service", () => {
        const service = services[6];
        const serviceSpec = generateServiceSpec(service);

        const result = developmentDockerComposeSpecFactory.create(
            serviceSpec, service
        );

        expect(result.services).toMatchSnapshot();
    });
});
