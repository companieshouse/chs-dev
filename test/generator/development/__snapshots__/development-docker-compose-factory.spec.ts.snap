// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevelopmentDockerComposeSpecFactory create adds builder service 1`] = `
{
  "service-seven": {
    "build": {
      "context": "/home/test-user/projects/docker-chs",
      "dockerfile": "locals/builder/java/v2/Dockerfile",
    },
    "depends_on": {
      "service-five": {
        "condition": "service_started",
      },
      "service-four": {
        "condition": "service_started",
      },
      "service-one": {
        "condition": "service_started",
      },
      "service-seven-builder": {
        "condition": "service_completed_successfully",
        "restart": true,
      },
      "service-six": {
        "condition": "service_started",
      },
      "service-three": {
        "condition": "service_started",
      },
      "service-two": {
        "condition": "service_started",
      },
    },
    "environment": [
      "ENV_VAR_ONE=one",
      "ENV_VAR_TWO=one",
    ],
    "expose": [
      8080,
    ],
    "healthcheck": {
      "interval": "60s",
      "test": "echo 1",
    },
    "image": "12345678.ecr.eu-west-2.amazonaws.com/local/service-seven:latest",
    "labels": [
      "chs.local.builder=repository",
      "chs.repository.url=git@github.com/companieshouse/repo-seven.git",
    ],
    "networks": [
      "chs",
    ],
    "volumes": [
      "repositories/service-seven/target:/opt/",
    ],
  },
  "service-seven-builder": {
    "build": {
      "context": "/home/test-user/projects/docker-chs",
      "dockerfile": "locals/builder/java/v2/build.Dockerfile",
    },
    "volumes": [
      "repositories/service-seven:/app/",
      "./out:/opt/out",
    ],
  },
}
`;
